packages/discovery-provider/src/tasks/index_payment_router.py

```diff
@@ -0,0 +1,858 @@
+            balance=Decimal(balance_change["post_balance"]),
+            tx_metadata=str(purchase_metadata["purchaser_user_id"]),
+        )
+        session.add(usdc_tx_received)
+        logger.debug(
+            f"index_payment_router.py | Creating usdc_tx_history received tx for purchase {usdc_tx_received}"
+        )
+
+
+def validate_and_index_usdc_transfers(
+    session: Session,
+    sender_account: str,
+    receiver_user_accounts: List[ReceiverUserAccount],
+    receiver_accounts: List[str],
+    balance_changes: dict[str, BalanceChange],
+    purchase_metadata: Union[PurchaseMetadataDict, None],
+    slot: int,
+    timestamp: datetime,
+    tx_sig: str,
+):
+    """Checks if the transaction is a valid purchase and if so creates the purchase record. Otherwise, indexes a transfer."""
+    if purchase_metadata is not None and validate_purchase(
+        purchase_metadata=purchase_metadata, balance_changes=balance_changes
+    ):
+        index_purchase(
+            session=session,
+            receiver_user_accounts=receiver_user_accounts,
+            receiver_accounts=receiver_accounts,
+            balance_changes=balance_changes,
+            purchase_metadata=purchase_metadata,
+            slot=slot,
+            timestamp=timestamp,
+            tx_sig=tx_sig,
+        )
+    # For invalid purchases or transfers not related to a purchase, we'll index
+    # it as a regular transfer, though it will always show as being sent from
+    # the payment router PDA
+    # TODO: We _could_ receive the actual sender from the first tranfer instruction here
+    # and use that instead.
+    else:
+        # TODO: Detect transfers _out_ of user banks to payment router and index them
+        # here as "sent" transactions
+        for user_account in receiver_user_accounts:
+            balance_change = balance_changes[user_account["user_bank_account"]]
+            usdc_tx_received = USDCTransactionsHistory(
+                user_bank=user_account["user_bank_account"],
+                slot=slot,
+                signature=tx_sig,
+                transaction_type=USDCTransactionType.transfer,
+                method=USDCTransactionMethod.receive,
+                transaction_created_at=timestamp,
+                change=Decimal(balance_change["change"]),
+                balance=Decimal(balance_change["post_balance"]),
+                tx_metadata=sender_account,
+            )
+            session.add(usdc_tx_received)
+            logger.debug(
+                f"index_payment_router.py | Creating transfer received tx {usdc_tx_received}"
+            )
+
+
+def process_route_instruction(
+    session: Session,
+    redis: Redis,
+    instruction: CompiledInstruction,
+    memos: List[str],
+    account_keys: List[str],
+    meta: UiTransactionStatusMeta,
+    tx_sig: str,
+    slot: int,
+    challenge_event_bus: ChallengeEventBus,
+    timestamp: datetime,
+):
+    # solana_client_manager: SolanaClientManager = (
```

Comment:
remove?

Reply:
Yeah I had left in some of the commented code thinking it would be easier to implement things later on. But I can always reference other indexer implementations when it comes to that.